Spring data JPA, can't find by FirstName

### Describe the bug


for a method like this:

findByLastNameContainingAndFirstNameContainingAndEmailContainingAllIgnoreCase

and we have this error:


### Expected behavior

It automatically implements the query

### Actual behavior

The given error is thrown:

Caused by: io.quarkus.spring.data.deployment.UnableToParseMethodException: Unable to parse query with limiting results clause.

### How to Reproduce?

1. Create a Quarkus Project: version 2.7.1

2. Follow the guide provided by quarkus on how to use the quarkus spring data extension

3. create an entity containing the field 
```java
String firstName;
```
4. try to compile

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

This line attempts to parse the field FirstName as a top like utility
(https://github.com/quarkusio/quarkus/blob/7aef8dfb215f2c7c75179de29f58d0c1d82368ec/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java#L99)

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java
@@ -94,33 +94,28 @@ public class MethodNameParser {
 
         // handle 'Top' and 'First'
         Integer topCount = null;
-        int firstIndex = methodName.indexOf("First");
-        int topIndex = methodName.indexOf("Top");
-        if ((firstIndex != -1) || (topIndex != -1)) {
-            int effectiveFirstOrTopIndex = (topIndex > -1 && firstIndex > -1) ? Math.max(topIndex, firstIndex)
-                    : Math.max(firstIndex, topIndex);
-            // 'First' and 'Top' could be part of a field name, so we only conside them as part of a top query
-            // if they are found before 'By'
-            if (effectiveFirstOrTopIndex < byIndex) {
-                try {
-                    String topCountStr = methodName.substring(effectiveFirstOrTopIndex, byIndex)
-                            .replace("Top", "").replace("First", "");
-                    if (topCountStr.isEmpty()) {
-                        topCount = 1;
-                    } else {
-                        topCount = Integer.valueOf(topCountStr);
-                    }
-                } catch (Exception e) {
-                    throw new UnableToParseMethodException(
-                            "Unable to parse query with limiting results clause. Offending method is "
-                                    + repositoryMethodDescription + ".");
-                }
+        int minFirstOrTopIndex = Math.min(indexOfOrMaxValue(methodName, "First"), indexOfOrMaxValue(methodName, "Top"));
+        // 'First' and 'Top' could be part of a field name, so we only conside them as part of a top query
+        // if they are found before 'By'
+        if (minFirstOrTopIndex < byIndex) {
+            if (queryType != QueryType.SELECT) {
+                throw new UnableToParseMethodException(
+                        "When 'Top' or 'First' is specified, the query must be a find query. Offending method is "
+                                + repositoryMethodDescription + ".");
             }
-        }
-        if ((topCount != null) && (queryType != QueryType.SELECT)) {
-            throw new UnableToParseMethodException(
-                    "When 'Top' or 'First' is specified, the query must be a find query. Offending method is "
-                            + repositoryMethodDescription + ".");
+            try {
+                String topCountStr = methodName.substring(minFirstOrTopIndex, byIndex)
+                        .replace("Top", "").replace("First", "");
+                if (topCountStr.isEmpty()) {
+                    topCount = 1;
+                } else {
+                    topCount = Integer.valueOf(topCountStr);
+                }
+            } catch (Exception e) {
+                throw new UnableToParseMethodException(
+                        "Unable to parse query with limiting results clause. Offending method is "
+                                + repositoryMethodDescription + ".");
+            }
         }
 
         if (methodName.substring(0, byIndex).contains("Distinct")) {
@@ -335,6 +330,11 @@ public class MethodNameParser {
         String whereQuery = where.toString().isEmpty() ? "" : " WHERE " + where.toString();
         return new Result(entityClass, "FROM " + getEntityName() + whereQuery, queryType, paramsCount, sort,
                 topCount);
+    }
+
+    private int indexOfOrMaxValue(String methodName, String term) {
+        int index = methodName.indexOf(term);
+        return index != -1 ? index : Integer.MAX_VALUE;
     }
 
     /**

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java
@@ -97,18 +97,24 @@ public class MethodNameParser {
         int firstIndex = methodName.indexOf("First");
         int topIndex = methodName.indexOf("Top");
         if ((firstIndex != -1) || (topIndex != -1)) {
-            try {
-                String topCountStr = methodName.substring(Math.max(firstIndex, topIndex), byIndex)
-                        .replace("Top", "").replace("First", "");
-                if (topCountStr.isEmpty()) {
-                    topCount = 1;
-                } else {
-                    topCount = Integer.valueOf(topCountStr);
+            int effectiveFirstOrTopIndex = (topIndex > -1 && firstIndex > -1) ? Math.max(topIndex, firstIndex)
+                    : Math.max(firstIndex, topIndex);
+            // 'First' and 'Top' could be part of a field name, so we only conside them as part of a top query
+            // if they are found before 'By'
+            if (effectiveFirstOrTopIndex < byIndex) {
+                try {
+                    String topCountStr = methodName.substring(effectiveFirstOrTopIndex, byIndex)
+                            .replace("Top", "").replace("First", "");
+                    if (topCountStr.isEmpty()) {
+                        topCount = 1;
+                    } else {
+                        topCount = Integer.valueOf(topCountStr);
+                    }
+                } catch (Exception e) {
+                    throw new UnableToParseMethodException(
+                            "Unable to parse query with limiting results clause. Offending method is "
+                                    + repositoryMethodDescription + ".");
                 }
-            } catch (Exception e) {
-                throw new UnableToParseMethodException(
-                        "Unable to parse query with limiting results clause. Offending method is "
-                                + repositoryMethodDescription + ".");
             }
         }
         if ((topCount != null) && (queryType != QueryType.SELECT)) {

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/Employee.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/Employee.java
@@ -20,6 +20,9 @@ public class Employee extends AbstractEntity {
     @Column(name = "last_name")
     private String lastName;
 
+    @Column(unique = true)
+    private String email;
+
     @ManyToOne(fetch = FetchType.LAZY)
     @JoinColumn(name = "team_id", nullable = false)
     private Team belongsToTeam;

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeRepository.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeRepository.java
@@ -7,4 +7,7 @@ import org.springframework.data.jpa.repository.JpaRepository;
 public interface EmployeeRepository extends JpaRepository<Employee, Long> {
 
     List<Employee> findByBelongsToTeamOrganizationalUnitName(String orgUnitName);
+
+    List<Employee> findByLastNameContainingAndFirstNameContainingAndEmailContainingAllIgnoreCase(String lastName,
+            String firstName, String emailPart);
 }

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeRepository.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeRepository.java
@@ -10,4 +10,6 @@ public interface EmployeeRepository extends JpaRepository<Employee, Long> {
 
     List<Employee> findByLastNameContainingAndFirstNameContainingAndEmailContainingAllIgnoreCase(String lastName,
             String firstName, String emailPart);
+
+    List<Employee> findFirst2ByFirstNameContainingIgnoreCaseOrderById(String firstName);
 }

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeResource.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeResource.java
@@ -42,4 +42,10 @@ public class EmployeeResource {
         return this.employeeRepository.findByLastNameContainingAndFirstNameContainingAndEmailContainingAllIgnoreCase(lastName,
                 firstName, email);
     }
+
+    @GET
+    @Path("/search-first-2")
+    public List<Employee> findTop2ByFirstNameContainingAllIgnoreCase(@QueryParam("first") String firstName) {
+        return this.employeeRepository.findFirst2ByFirstNameContainingIgnoreCaseOrderById(firstName);
+    }
 }

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeResource.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/EmployeeResource.java
@@ -6,6 +6,7 @@ import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
 import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
 
 @Path("/employee")
 @Produces("application/json")
@@ -32,5 +33,13 @@ public class EmployeeResource {
     @Path("/unit/{orgUnitName}")
     public List<Employee> findByManagerOfManager(@PathParam("orgUnitName") String orgUnitName) {
         return this.employeeRepository.findByBelongsToTeamOrganizationalUnitName(orgUnitName);
+    }
+
+    @GET
+    @Path("/search")
+    public List<Employee> findByLastNameContainingAndFirstNameContainingAndEmailContainingAllIgnoreCase(
+            @QueryParam("first") String firstName, @QueryParam("last") String lastName, @QueryParam("email") String email) {
+        return this.employeeRepository.findByLastNameContainingAndFirstNameContainingAndEmailContainingAllIgnoreCase(lastName,
+                firstName, email);
     }
 }

--- a/integration-tests/spring-data-jpa/src/main/resources/import.sql
+++ b/integration-tests/spring-data-jpa/src/main/resources/import.sql
@@ -80,7 +80,8 @@ INSERT INTO team(id, name, unit_id) VALUES (11, 'Sales Team', 2);
 INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (100, 'johdoe', 'John', 'Doe', 10, 'johdoe@acme.com');
 INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (101, 'petdig', 'Peter', 'Digger', 10, 'petdig@acme.com');
 INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (102, 'stesmi', 'Stella', 'Smith', 11, 'stesmi@acme.com');
-INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (100, 'jd', 'John', 'Doe', 12, 'jd@acme.com');
+INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (103, 'jd', 'John', 'Doe', 11, 'jd@acme.com');
+INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (104, 'jodo', 'John', 'Doe', 11, 'jodo@acme.com');
 
 INSERT INTO MotorCar(id, brand, model) VALUES (1, 'Monteverdi', 'Hai 450');
 INSERT INTO MotorCar(id, brand, model) VALUES (2, 'Rinspeed', 'iChange');

--- a/integration-tests/spring-data-jpa/src/main/resources/import.sql
+++ b/integration-tests/spring-data-jpa/src/main/resources/import.sql
@@ -77,9 +77,10 @@ INSERT INTO unit(id, name) VALUES (1, 'Delivery Unit');
 INSERT INTO unit(id, name) VALUES (2, 'Sales and Marketing Unit');
 INSERT INTO team(id, name, unit_id) VALUES (10, 'Development Team', 1);
 INSERT INTO team(id, name, unit_id) VALUES (11, 'Sales Team', 2);
-INSERT INTO employee(id, user_id, first_name, last_name, team_id) VALUES (100, 'johdoe', 'John', 'Doe', 10);
-INSERT INTO employee(id, user_id, first_name, last_name, team_id) VALUES (101, 'petdig', 'Peter', 'Digger', 10);
-INSERT INTO employee(id, user_id, first_name, last_name, team_id) VALUES (102, 'stesmi', 'Stella', 'Smith', 11);
+INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (100, 'johdoe', 'John', 'Doe', 10, 'johdoe@acme.com');
+INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (101, 'petdig', 'Peter', 'Digger', 10, 'petdig@acme.com');
+INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (102, 'stesmi', 'Stella', 'Smith', 11, 'stesmi@acme.com');
+INSERT INTO employee(id, user_id, first_name, last_name, team_id, email) VALUES (100, 'jd', 'John', 'Doe', 12, 'jd@acme.com');
 
 INSERT INTO MotorCar(id, brand, model) VALUES (1, 'Monteverdi', 'Hai 450');
 INSERT INTO MotorCar(id, brand, model) VALUES (2, 'Rinspeed', 'iChange');

--- a/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/EmployeeResourceTest.java
+++ b/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/EmployeeResourceTest.java
@@ -30,4 +30,12 @@ public class EmployeeResourceTest {
         assertThat(employees).extracting("userId").containsExactly("johdoe");
     }
 
+    @Test
+    public void testFindFirst2ByFirstNameContainingIgnoreCase() {
+        List<Employee> employees = when().get("/employee/search-first-2?first=John").then()
+                .statusCode(200)
+                .extract().body().jsonPath().getList(".", Employee.class);
+
+        assertThat(employees).extracting("userId").containsExactly("johdoe", "jd");
+    }
 }

--- a/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/EmployeeResourceTest.java
+++ b/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/EmployeeResourceTest.java
@@ -21,4 +21,13 @@ public class EmployeeResourceTest {
         assertThat(employees).extracting("userId").containsExactlyInAnyOrder("johdoe", "petdig");
     }
 
+    @Test
+    public void testFindByLastNameContainingAndFirstNameContainingAndEmailContainingAllIgnoreCase() {
+        List<Employee> employees = when().get("/employee/search?first=John&last=Doe&email=doe").then()
+                .statusCode(200)
+                .extract().body().jsonPath().getList(".", Employee.class);
+
+        assertThat(employees).extracting("userId").containsExactly("johdoe");
+    }
+
 }

